File: backend/src/schemas/audience.py
"""Audience request and response schemas.

Only ``full_name`` is ever required of a member. Everything else is optional
because a real uploaded list is ragged, and a list that refuses a row for a
missing phone number is a list nobody can upload.
"""

from datetime import datetime

from pydantic import EmailStr, Field, field_validator

from src.schemas.common import StrictModel
from src.schemas.enums import AgeGroup, Gender
from src.schemas.validators import clean_text, normalise_place

# A phone as E.164 without separators. Matches what the importer normalises to.
PHONE_PATTERN = r"^\+?[1-9]\d{6,19}$"

# How many members one write may carry. A bigger file is imported in batches.
MAX_MEMBERS_PER_WRITE = 1000


class AudienceMemberInput(StrictModel):
    full_name: str = Field(
        ...,
        min_length=1,
        max_length=80,
        description="The only required field. Resolves {{customer_name}} in campaign copy.",
    )
    email: EmailStr | None = None
    phone: str | None = Field(None, max_length=20, pattern=PHONE_PATTERN)
    age: int | None = Field(
        None,
        ge=13,
        le=120,
        description="The age itself, not a bracket. Brackets are derived at read time.",
    )
    gender: Gender = Gender.UNKNOWN
    city: str | None = Field(None, max_length=80)
    country: str | None = Field(None, max_length=56)
    external_ref: str | None = Field(None, max_length=120, description="CRM contact id.")
    attributes: dict[str, str] | None = Field(
        None, description="Free-form extras carried through import and export untouched."
    )

    @field_validator("full_name", "external_ref")
    @classmethod
    def _clean(cls, value: str | None) -> str | None:
        return clean_text(value)

    @field_validator("city", "country")
    @classmethod
    def _place(cls, value: str | None) -> str | None:
        return normalise_place(value)


class AudienceMemberRead(StrictModel):
    id: str
    full_name: str
    email: str | None = None
    phone: str | None = None
    age: int | None = None
    age_group: AgeGroup = Field(
        AgeGroup.UNKNOWN, description="Derived from age. Never stored, so it cannot go stale."
    )
    gender: Gender = Gender.UNKNOWN
    city: str | None = None
    country: str | None = None
    external_ref: str | None = None
    attributes: dict[str, str] | None = None
    created_at: datetime


class AudienceMemberPage(StrictModel):
    """One filtered page of people, plus how many matched the filter."""

    items: list[AudienceMemberRead]
    total: int
    limit: int
    offset: int


class SegmentBucket(StrictModel):
    """One slice of a breakdown.

    ``key`` is an enum value for age and gender and the stored text for a city
    or a country. The frontend owns the human label either way.
    """

    key: str
    count: int
    share: float = Field(..., description="Fraction of the whole audience, 0-1, rounded to 4dp.")


class AudienceSegments(StrictModel):
    """What the audience is made of, without naming anybody.

    This is the screen's headline. A list of 100 names tells you nothing; the
    same 100 people broken down by age, gender and place is what you choose a
    campaign's targeting from.
    """

    total: int
    with_email: int = Field(..., description="Members reachable by email.")
    with_phone: int = Field(..., description="Members reachable by phone.")
    age_groups: list[SegmentBucket] = Field(default_factory=list)
    genders: list[SegmentBucket] = Field(default_factory=list)
    cities: list[SegmentBucket] = Field(default_factory=list)
    countries: list[SegmentBucket] = Field(default_factory=list)


class AudienceCreate(StrictModel):
    name: str = Field(..., min_length=1, max_length=120)
    description: str | None = Field(None, max_length=500)
    members: list[AudienceMemberInput] = Field(
        default_factory=list, max_length=MAX_MEMBERS_PER_WRITE
    )

    @field_validator("name", "description")
    @classmethod
    def _clean(cls, value: str | None) -> str | None:
        return clean_text(value)


class AudienceUpdate(StrictModel):
    """Renames and re-describes. Membership is managed on its own routes."""

    name: str | None = Field(None, min_length=1, max_length=120)
    description: str | None = Field(None, max_length=500)

    @field_validator("name", "description")
    @classmethod
    def _clean(cls, value: str | None) -> str | None:
        return clean_text(value)


class AudienceMembersInput(StrictModel):
    """Body for a bulk add. One CSV upload is one call."""

    members: list[AudienceMemberInput] = Field(..., min_length=1, max_length=MAX_MEMBERS_PER_WRITE)


class SkippedMember(StrictModel):
    """One row that did not land, and why."""

    index: int = Field(..., description="Position in the submitted list, 0-based.")
    full_name: str
    reason: str


class AudienceImportResult(StrictModel):
    """The outcome of a bulk add.

    A partial success is the normal case, not an error: one repeated email in a
    200-row file should cost that row, not the file. Every skipped row is named
    so the user can see what did not land instead of silently losing it.
    """

    added: int
    skipped: list[SkippedMember] = Field(default_factory=list)
    member_count: int = Field(..., description="Size of the audience after the import.")


class AudienceListItem(StrictModel):
    id: str
    name: str
    description: str | None = None
    member_count: int = 0
    campaign_count: int = Field(0, description="Campaigns currently pointing at this audience.")
    created_at: datetime
    updated_at: datetime


class AudienceRead(AudienceListItem):
    """One audience with its breakdown. Members come from the members route."""

    segments: AudienceSegments


class AudiencePage(StrictModel):
    items: list[AudienceListItem]
    total: int
    limit: int
    offset: int


class AudienceSelection(StrictModel):
    """What a campaign says about the list it targets."""

    id: str
    name: str
    member_count: int

File: frontend/src/types/audience.ts
/**
 * The audience: a reusable, named list of people that campaigns select.
 *
 * Owned by the account rather than by one campaign, which is what lets the
 * same list run several of them. Mirrors `backend/src/schemas/audience.py`.
 */

export type Gender = "FEMALE" | "MALE" | "OTHER" | "UNKNOWN";

export type AgeGroup =
  | "UNDER_18"
  | "AGE_18_24"
  | "AGE_25_34"
  | "AGE_35_44"
  | "AGE_45_54"
  | "AGE_55_64"
  | "AGE_65_PLUS"
  | "UNKNOWN";

/**
 * One person.
 *
 * Only `full_name` is ever guaranteed. Everything else is nullable because an
 * uploaded list is ragged, and the screens have to render a row that is a name
 * and nothing else.
 */
export interface AudienceMember {
  id: string;
  /** Resolves `{{customer_name}}` in campaign copy. */
  full_name: string;
  email: string | null;
  phone: string | null;
  age: number | null;
  /** Derived server-side from `age`, so it cannot go stale on a birthday. */
  age_group: AgeGroup;
  gender: Gender;
  city: string | null;
  country: string | null;
  /** CRM contact id. */
  external_ref: string | null;
  attributes: Record<string, string> | null;
  created_at: string;
}

export interface AudienceMemberPage {
  items: AudienceMember[];
  /** Counts what matched the filter, not the whole list. */
  total: number;
  limit: number;
  offset: number;
}

/** One slice of a breakdown. `key` is an enum value, or the place itself. */
export interface SegmentBucket {
  key: string;
  count: number;
  /** Fraction of the whole audience, 0-1. */
  share: number;
}

/** What the audience is made of, without naming anybody. */
export interface AudienceSegments {
  total: number;
  with_email: number;
  with_phone: number;
  age_groups: SegmentBucket[];
  genders: SegmentBucket[];
  cities: SegmentBucket[];
  countries: SegmentBucket[];
}

export interface AudienceSummary {
  id: string;
  name: string;
  description: string | null;
  member_count: number;
  /** Campaigns currently pointing at this audience. Blocks deletion when > 0. */
  campaign_count: number;
  created_at: string;
  updated_at: string;
}

/** One audience with its breakdown. Members come from the members endpoint. */
export interface Audience extends AudienceSummary {
  segments: AudienceSegments;
}

/**
 * What a campaign says about the list it targets.
 *
 * Deliberately thinner than {@link AudienceSummary}: a campaign read carries
 * only enough to name the list and say how big it is. Everything else about
 * the audience is fetched from the audience itself.
 */
export interface AudienceSelection {
  id: string;
  name: string;
  member_count: number;
}

export interface AudiencePage {
  items: AudienceSummary[];
  total: number;
  limit: number;
  offset: number;
}

/* -------------------------------------------------------------------------
 * Write payloads
 * ---------------------------------------------------------------------- */

/**
 * One person as the API accepts them.
 *
 * Every field but the name is optional, and that is the contract the CSV
 * importer is built against: a file with one column still uploads.
 */
export interface AudienceMemberInput {
  full_name: string;
  email?: string | null;
  phone?: string | null;
  age?: number | null;
  gender?: Gender;
  city?: string | null;
  country?: string | null;
  external_ref?: string | null;
}

export interface AudienceWritePayload {
  name: string;
  description?: string | null;
  members?: AudienceMemberInput[];
}

export type AudienceUpdatePayload = Partial<Pick<AudienceWritePayload, "name" | "description">>;

/** One row that did not land, and why. */
export interface SkippedMember {
  index: number;
  full_name: string;
  reason: string;
}

/**
 * The outcome of a bulk add.
 *
 * A partial success is the normal case: one repeated email costs its row, not
 * the file, and every row that did not land is named.
 */
export interface AudienceImportResult {
  added: number;
  skipped: SkippedMember[];
  /** Size of the audience after the import. */
  member_count: number;
}

/* -------------------------------------------------------------------------
 * Labels
 *
 * The API stores enum values and the UI owns the words, the same split the
 * CTA and status enums use.
 * ---------------------------------------------------------------------- */

export const AGE_GROUP_LABELS: Record<AgeGroup, string> = {
  UNDER_18: "Under 18",
  AGE_18_24: "18–24",
  AGE_25_34: "25–34",
  AGE_35_44: "35–44",
  AGE_45_54: "45–54",
  AGE_55_64: "55–64",
  AGE_65_PLUS: "65+",
  UNKNOWN: "Not given",
};

/** The order a breakdown reads in, which is by age and not by size. */
export const AGE_GROUP_ORDER: AgeGroup[] = [
  "UNDER_18",
  "AGE_18_24",
  "AGE_25_34",
  "AGE_35_44",
  "AGE_45_54",
  "AGE_55_64",
  "AGE_65_PLUS",
  "UNKNOWN",
];

export const GENDER_LABELS: Record<Gender, string> = {
  FEMALE: "Female",
  MALE: "Male",
  OTHER: "Other",
  UNKNOWN: "Not given",
};

export const GENDER_ORDER: Gender[] = ["FEMALE", "MALE", "OTHER", "UNKNOWN"];

/** Label a bucket key, falling back to the key for cities and countries. */
export function segmentLabel(key: string): string {
  return (
    AGE_GROUP_LABELS[key as AgeGroup] ??
    GENDER_LABELS[key as Gender] ??
    (key === "UNKNOWN" ? "Not given" : key)
  );
}
